proposals/lip-0049.md

```diff
@@ -223,135 +223,171 @@ When processing a cross-chain message `ccm`, the functions below may be called b
+            typeID = TYPE_ID_CCM_PROCESSED,
+            data = {"ccmID": ccmID, "status": CCM_PROCESSED_STATUS_COMMAND_DISCARDED},
+            topics = [ccm.sendingChainID, ccm.receivingChainID]
+        )
 ```
 
 #### forward
 
-The `forward` function forwards a CCM to the specified receiving chain. This function is only called on the mainchain when processing a CCU.
+The `forward` function forwards a CCM to the specified receiving chain. This function is only called on the mainchain when processing a CCU. If an error occurs, the CCM is bounced back to the sending chain or discarded.
 
 ```python
-# The ccu including this message is passed as context to this function
-forward(ccu, ccm):
-    if (chainAccount(ccm.receivingChainID) exists
-        and chainAccount(ccm.receivingChainID).status == CHAIN_ACTIVE
-        and isLive(ccm.receivingChainID):
-        # The Token module is called to transfer the fee from sending to receiving chain
-        token.forwardMessageFee(ccm)
-        if the previous call fails or returns False:
-            return
+def forward(ccu: CCU, ccm: CCM) -> None:
+    # Calculate CCM ID, used later in events.
+    ccmID = SHA-256(encode(crossChainMessageSchema, ccm))
+    
+    # Chain with status REGISTERED do not pass this check.
+    if isLive(ccm.receivingChainID) and chainAccount(ccm.receivingChainID).status == CHAIN_STATUS_ACTIVE:
+        # Transfer the fee from escrow to escrow, call Token LIP
+        Token.transferBetweenEscrows(ccm.sendingChainID, ccm.receivingChainID, feeTokenID, fee)
+
         addToOutbox(ccm.receivingChainID, ccm)
+        # Emit CCM forwarded event.
+        emitEvent(
+            moduleID = MODULE_ID_INTEROPERABILITY,
+            typeID = TYPE_ID_CCM_PROCESSED,
+            data = {"ccmID": ccmID, "status": CCM_PROCESSED_STATUS_FORWARDED},
+            topics = [ccm.sendingChainID, ccm.receivingChainID]
+        )
         return
 
+    # If ccm status is not ok, stop the processing of this message
     if ccm.status != CCM_STATUS_OK:
-        # Stop the processing of this message
+        # Emit CCM discarded event.
+        emitEvent(
+            moduleID = MODULE_ID_INTEROPERABILITY,
+            typeID = TYPE_ID_CCM_PROCESSED,
+            data = {"ccmID": ccmID, "status": CCM_PROCESSED_STATUS_DISCARDED},
+            topics = [ccm.sendingChainID, ccm.receivingChainID]
+        )
         return
 
     # In all following cases, bounce the message with updated status
     # Note: the fee was not assigned to the mainchain relayer and is also bounced
-    bounce(ccm)
+    
+    # Emit CCM processed event: channel unavailable.
+    emitEvent(
+        moduleID = MODULE_ID_INTEROPERABILITY,
+        typeID = TYPE_ID_CCM_PROCESSED,
+        data = {"ccmID": ccmID, "status": CCM_PROCESSED_STATUS_CHANNEL_UNAVAILABLE},
+        topics = [ccm.sendingChainID, ccm.receivingChainID]
+    )
+    ccm.status = CCM_STATUS_CHANNEL_UNAVAILABLE
+    addToOutbox(ccm.sendingChainID, ccm)
```

Comment:
Why do I not use the `send` method here, doing basically the same as in the `error` method with a different fee.

Reply:
We do not want to increase the `nonce` and send it from the mainchain

Reply:
As we now use `bounce` in these cases, this functionality is now cleanly encapsulated. All good.